docs/designs/data_transform.md

```diff
@@ -1,4 +1,84 @@
+
+Currently, SQLFlow converts content in the COLUMN clause into the Python source code of data transformation.  The following example would call the Feature Column API [`tf/feature_column/categorical_column_with_hash_bucket`](https://www.tensorflow.org/api_docs/python/tf/feature_column/categorical_column_with_hash_bucket) from TensorFlow to convert a string `address` from the database into a single-element tensor of an integer ID.
+
+```sql
+SELECT * FROM welfare TO TRAIN DNNRegressor COLUMN hash_bucket(address, 1000) LABEL income;
+```
+
+However, data transform often requires parameters. The above example requires the bucket size, 1000. Other examples include
+standardize(age, mean, stdev)
+normalize(size, min, max)
+categorize(city, city_name_list)
+
+The challenge is -- users don't want to specify these parameters; instead, they want SQLFlow does the data statistics and derives these parameters from data automatically.  So, the above examples become
+standardize(age)
+normalize(size)
+categorize(city)
+
+In the terminology of TFX Transform, the word *analysis* refers to data statistics that derives the above parameters, given which, the *transforming* refers to the conversion of raw data into model inputs.  This design document is about making SQLFlow support data analysis, in addition to transformation.
+
+
+## The Syntax Extension
+
+We can ask users to write SQL statements for data analysis.  For example, we might want to assume that the following code snippet trains a model to predict the plate price from its size.
+
+```sql
+min, max = SELECT min(age), max(age) FROM plates;
+SELECT * FROM plates TO TRAIN DNNRegressor COLUMN normalize(size, min, max) LABEL price;
+```
+
+Unfortunately, SQL doesn't provide variables. The above example changes so much to the syntax and semantics that most SQL users cannot accept it.
+
+Also, to fulfill the goal of making deep learning more straightforward, we hope our users can write the following statement.
+
+```sql
+SELECT * FROM plates TO TRAIN DNNRegressor COLUMN normalize(size) LABEL price INTO a_model;
+```
+
+This syntax implies that SQLFlow can convert the above statement into a Python program that does data analysis and data transformation.
+
+To use the trained model to make predictions, we hope the users don't have to rewrite the COLUMN clause.   For example, the following statement assumes that the table `new_plates` has a field `size`. The converted Python code should normalize `size` before sending it to the input of `a_mode`.
+
+```sql
+SELECT * FROM new_plates TO PREDICT price USING a_model;
+```
+
+If the user doesn't want normalization, but the raw value, as the input, she could write the following.
+
+```sql
+SELECT * FROM new_plate TO PREDICT price USING a_model COLUMN size
```

Comment:
I'm afraid that the Prediction job will fail if use can write COLUMN in the `TO PREDICT` expression.
Let's take iris model for example:
If we train the model using
```SQL
SELECT * FROM iris.train
TO TRAIN DNNClassifer
WITH hidden_units = [10, 10], n_classes = 3, EPOCHS = 10
COLUMN sepal_length, sepal_width, petal_length, petal_width
LABEL class
INTO sqlflow_models.my_dnn_model;
```
And make prediction using:
```
SELECT * FROM iris.test
TO PREDICT iris.predict.class
USING sqlflow.my_dnn_model
COLUMN sepal_length, sepal_width, petal_length
```
The prediction job will fail. The reason is the expected input of the trained model contains four elements. But the actual input contains three.

Reply:
And this failure is expected in our design. SQLFlow should prompt a message like "The inputs to the model during prediction are not compatible with the model specification.".  So, I think it is not a problem to allow a COLUMN clause in TO PREDICT.